### PR TITLE
Refactor QPDFObject creation and cloning

### PR DESCRIPTION
**Pull request by m-holger**
Move responsibility for creating shared pointers to objects and cloning from QPDFObjectHandle to QPDFObject.

### CODE REVIEW

**Comment by jberkenbilt** [comment]
I think I like this. I need to think through the use of the weak pointer. I'm definitely comfortable with the concept of weak pointers but have never used them in qpdf since, as you know, I was using PointerHolder until the current unreleased changes. The switch to std::shared_ptr makes this possible.

I'd also like to think through the feasibility of changing QPDF_Array and QPDF_Dictionary to hang onto shared_ptr<QPDFObject> instead of QPDFObjectHandle and whether this is a step along the path to doing that. I wonder whether having a parent pointer in QPDFObject as a weak pointer might work as a way to eventually eliminate the shared direct object problem while detecting when it happens. If the situation isn't too bad, I might be up for making it a warning in qpdf 11 and making it a breaking change in qpdf 12.

The method QPDFObjectHandle::releaseResolved breaks circular references by explicitly clearing object pointers in QPDFObjectHandle at destruction. It took a few tries to get that logic right. I know there are tests that fail if that step is missed, though they may only fail when memory sanitizers are in use since they create leaks. I'm pretty sure that code will need to be retained even with weak pointers since it comes about whenever a circular reference exists with actual PDF objects. Object cycles are normal in correct, compliant PDF files. I still would need to think through the interaction of that with this, if any.

**Comment by m-holger** [comment]
> I'd also like to think through the feasibility of changing QPDF_Array and QPDF_Dictionary to hang onto shared_ptr 

I assume you mean shared_ptr<QPDFObject>. If so, there are a couple of implications
QPDFObject

- qpdf, objid, and generation would need to be moved from QPDFObjectHandle to QPDFObject
- shallowCopy will need to make proper copies of scalar objects
- the way unreserved objects are handled probably would need to change (I haven't checked

**Comment by m-holger** [comment]
Sorry, accidentally closed

File: qpdf/qpdf.testcov [hunk 1]
```diff
@@ -70,12 +70,12 @@ QPDFTokenizer null in name 0
 QPDFTokenizer bad name 0
 QPDF_Stream invalid filter 0
 QPDF UseOutlines but no Outlines 0
-QPDFObjectHandle clone bool 0
```
Comment (jberkenbilt) [inline qpdf/qpdf.testcov:73]: Just remove these without replacing them.

File: qpdf/qpdf.testcov [hunk 2]
```diff
@@ -199,8 +199,8 @@ QPDF updateAllPagesCache 0
 QPDF insert non-indirect page 0
 QPDF insert indirect page 0
 QPDFObjectHandle ERR shallow copy stream 0
-QPDFObjectHandle shallow copy array 0
```
Comment (jberkenbilt) [inline qpdf/qpdf.testcov:202]: just remove

**Review by jberkenbilt — CHANGES_REQUESTED** [review]
First, I really like this change. There are as few things that need to be different though.

There are two separate things going on, both involving moving things into the QPDFObject classes.

* create -- This is a great change as is.

* clone -- Let's replace `clone()` with `shallowCopy()`. Otherwise, it looks great.

I don't want a method called `clone()` because it is ambiguous as to what kind of copy it is. Perhaps you chose this based on my use of the term `clone` in the coverage cases, but I only used it for types where there is no distinction between a deep and shallow copy.

By the way, my convention is to put the name of the class with the coverage case as the beginning of the second argument to QTC::TC, so all the places where you had `QTC::TC("qpdf", "QPDFObject ...")` should have been `QTC::TC("qpdf", "QPDF_Bool ...")`, etc. It's possible that there may be exceptions to this, and I know there's grep, but it just helps me at a glance when I'm looking at coverage cases.

That said, it has recently come to my attention that `QTC::TC` is slow on Windows if there is a big environment, and while I plan to do some optimization, I'm going to remove coverage cases in performance critical or commonly called sections of code. Some of these were added really early on, and there's just no way the copying code is not covered, so the explicit coverage cases don't add any value. I'm planning on removing a bunch, including those ones, so you may as well just drop the coverage calls.

Thanks for this -- this is a great improvement.

File: libqpdf/QPDFObject.cc [hunk 3]
```diff
@@ -6,6 +6,14 @@ QPDFObject::QPDFObject() :
 {
 }
 
+std::shared_ptr<QPDFObject>
+QPDFObject::do_create(QPDFObject* object)
+{
+    std::shared_ptr<QPDFObject> obj(object);
+    obj->self = obj;
```
[the remark this answers is not in the document]
Comment (m-holger) [reply libqpdf/QPDFObject.cc]: Removed for now

File: libqpdf/QPDF_Array.cc [hunk 4]
```diff
@@ -1,6 +1,7 @@
 #include <qpdf/QPDF_Array.hh>
 
 #include <qpdf/QIntC.hh>
+#include <qpdf/QTC.hh>
```
[the remark this answers is not in the document]
Comment (m-holger) [reply libqpdf/QPDF_Array.cc]: Done

File: libqpdf/QPDF_Bool.cc [hunk 5]
```diff
@@ -1,10 +1,25 @@
 #include <qpdf/QPDF_Bool.hh>
 
+#include <qpdf/QTC.hh>
```
[the remark this answers is not in the document]
Comment (m-holger) [reply libqpdf/QPDF_Bool.cc]: Done

File: libqpdf/QPDF_Dictionary.cc [hunk 6]
```diff
@@ -2,13 +2,27 @@
...
+std::shared_ptr<QPDFObject>
+QPDF_Dictionary::create(std::map<std::string, QPDFObjectHandle> const& items)
+{
+    return do_create(new QPDF_Dictionary(items));
+}
+
+std::shared_ptr<QPDFObject>
+QPDF_Dictionary::clone()
+{
+    QTC::TC("qpdf", "QPDFObject shallow copy dictionary");
```
[the remark this answers is not in the document]
Comment (m-holger) [reply libqpdf/QPDF_Dictionary.cc]: Done

File: qpdf/qpdf.testcov (hunk 1 above)
Comment (m-holger) [reply qpdf/qpdf.testcov:73]: Done

File: qpdf/qpdf.testcov (hunk 2 above)
Comment (m-holger) [reply qpdf/qpdf.testcov:202]: Done

File: libqpdf/QPDF_Bool.cc [hunk 7]
```diff
@@ -1,10 +1,25 @@
...
+std::shared_ptr<QPDFObject>
+QPDF_Bool::create(bool value)
+{
+    return do_create(new QPDF_Bool(value));
+}
+
+std::shared_ptr<QPDFObject>
+QPDF_Bool::clone()
+{
+    QTC::TC("qpdf", "QPDFObject clone bool");
```
[the remark this answers is not in the document]
Comment (m-holger) [reply libqpdf/QPDF_Bool.cc]: Done

File: libqpdf/QPDF_Integer.cc [hunk 8]
```diff
@@ -1,12 +1,26 @@
 #include <qpdf/QPDF_Integer.hh>
 
+#include <qpdf/QTC.hh>
```
[the remark this answers is not in the document]
Comment (m-holger) [reply libqpdf/QPDF_Integer.cc]: Done

**Comment by jberkenbilt** [comment]
Great except you called it `shallow_copy` instead of `shallowCopy`. For C++, cameCase is the convention for method names and is used for everything else in QPDFObject and QPDFObjectHandle, so I would prefer `shallowCopy`. Otherwise, I think this is ready to go.

**Comment by m-holger** [comment]
I plead temporary insanity!

**Comment by jberkenbilt** [comment]
Thanks for this change! (I'm going to accept `do_create` rather than `doCreate` on the basis of its being a static method.)

**Pull request merged**